### ISSUE TITLE
Invoking tests against editable installs have been masking bugs

### ISSUE THREAD

**Issue by jacobtylerwalls** [pylint-dev/pylint#8854]
We've been invoking tests on CI against an editable install of pylint for a long time. Doing so has the potential to mask issues. See [blog post](https://blog.ganssle.io/articles/2019/08/test-as-installed.html).

We now have evidence of some of those issues from the [failures](https://github.com/pylint-dev/pylint/actions/runs/5556300913/jobs/10148631043) in the PR testing against Python 3.12, which for unrelated reasons, forced us to upgrade `setuptools`. Without using the [legacy behavior](https://setuptools.pypa.io/en/latest/userguide/development_mode.html#legacy-behavior), bugs became evident.

One of the most minimal bugs is this one:

https://github.com/pylint-dev/pylint/blob/9f2de9123b579fadd512e31e9b68c7eb40a9c62d/tests/lint/test_pylinter.py#L39-L41

That mock shouldn't work, since the real location of the `PyLinter` class is at `pylint.lint.pylinter.PyLinter`. It was only working because the `pylint.lint` module imports `PyLinter`, but AFIAU there's no reason to expect that to work when testing the installation of the `pylint` package, which won't install `pylint.lint` (whereas an editable install will, assuming you're invoking tests from `pylint`.

That's just a bug in a test, but there are functional bugs, too, such as the moved check in #8702 potentially crashing:

```python
File "/Users/<user>/pylint/pylint/config/config_initialization.py", line 114, in _config_initialization
    for exc_name in linter.config.overgeneral_exceptions:
AttributeError: 'Namespace' object has no attribute 'overgeneral_exceptions'
```

***

I've found a workaround for the 3.12 branch to keep that one using legacy behavior, but suggesting we also look at this fixing this for the next pylint alpha.

**Comment by jacobtylerwalls** [comment]
This is no longer as crash-y as before. I haven't worked out when or why, but it's still worth doing.

**Comment by jacobtylerwalls** [comment]
> This is no longer as crash-y as before. I haven't worked out when or why
